README.md
# Netflix Login

[![Circle CI](https://circleci.com/gh/johnf/netflix-login-node.svg?style=svg)](https://circleci.com/gh/johnf/netflix-login-node)
[![Coverage Status](https://coveralls.io/repos/johnf/netflix-login-node/badge.svg?branch=master&service=github)](https://coveralls.io/github/johnf/netflix-login-node?branch=master)

Login in to Netflix and provide cookies and encryption keys.

## Usage Notes

Please read the [Netflix Terms of Use](https://help.netflix.com/legal/termsofuse?locale=en&docType=termsofuse) before using this module.

## Installation

``` bash
npm install netflix-login
```

## Usage

``` javascript

import netflixLogin from 'netflix-login/login';
import netflixCrypto from 'netflix-login/crypto';

const options = {
  useCache: true,
  cachePath: 'tmp',
};
const username = 'johnf@inodes.org';
const password = 'secret';

netflixLogin.login(username, password, options)
  .then(authData => netflixCrypto.fetchCryptoKeys(authData, options))
  .then(data => console.error(data))
  .catch((error) => {
    console.error('ERROR');
    console.error(error);
  });
```

## Development

After checking out the repo, run `npm test` to run the tests.

To release a new version:

* npm test
* npm version major|minor|patch
* npm publish

This will run the tests, update the version, create a git tag for the version, push git commits and tags. Publish the module file to [npmjs.com](https://npmjs.com).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/johnf/netflix-login-node. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [Contributor Covenant](contributor-covenant.org) code of conduct.

## License

The gem is available as open source under the terms of the [ISC License](http://opensource.org/licenses/ISC).


CHANGELOG.md
# Changelog

## 1.1.0 - 2017-02-25

* dist dir was missing (@johnf)

## 1.0.0 - 2017-02-25

* Convert to ES6 (@johnf)

## 0.0.4

* Cache credentials to file for reuse (@johnf)
* Generate crypto keys (@johnf)

## 0.0.3

* Add Netflix Terms of Use to README.md (@johnf)

## 0.0.2

* First release (@johnf)


test/index.js
"use strict";

var expect = require('chai').expect;
var nock = require('nock');

var netflixLogin = require('../index');

var Purdy = require('purdy');

// Setup nock
nock.disableNetConnect();
// nock.enableNetConnect();
// nock.recorder.rec();

var username = 'johnf@inodes.org';
var password = 'secret';
var authURL = 'authURL=1468047668683.authURL';
var netflixId = 'netflixCookie';
var secureNetflixId = 'secureNetflixCookie';
var esn = 'ESN';

var nockOptions = {
  allowUnmocked: false
};

// getAuthURL
nock('https://www.netflix.com', nockOptions)
  .get('/Login')
  .reply(302, '', { location: 'https://www.netflix.com/gb/Login' });

nock('https://www.netflix.com', nockOptions)
  .get('/gb/Login')
  .reply(200, 'name="authURL" value="' + authURL + '"');

// postLogin
var body = 'authURL=' + encodeURIComponent(authURL) +
  '&email=' + encodeURIComponent(username) +
  '&password=' + password +
  '&rememberMeCheckbox=true' +
  '&flow=websiteSignUp' +
  '&mode=login' +
  '&action=loginAction' +
  '&withFields=email%2Cpassword%2CrememberMe%2CnextPage' +
  '&nextPage=';
nock('https://www.netflix.com')
  .post('/Login', body)
  .reply(302, '', {
    location: 'https://www.netflix.com/gb/',
    'set-cookie': [
      'NetflixId=' + netflixId + '; Domain=.netflix.com; Path=/; Expires=Sun, 09 Jul 2017 12:44:54 GMT; HttpOnly',
      'SecureNetflixId=' + secureNetflixId + '; Domain=.netflix.com; Path=/; Expires=Fri, 01 Jan 2038 00:00:00 GMT; HttpOnly; Secure',
    ],
  });

// getLogin
nock('https://www.netflix.com', nockOptions)
  .get('/Login')
  .reply(302, '', { location: 'https://www.netflix.com/gb/Login' });

nock('https://www.netflix.com', nockOptions)
  .get('/gb/Login')
  .reply(302, '', { location: 'https://www.netflix.com/browse', });

nock('https://www.netflix.com', nockOptions)
  .get('/browse')
  .reply(200, '"esn":"' + esn + '"');

describe('Netflix Login', function() {

  beforeEach(function resetNetflix() {
    netflixLogin._reset();
  });

  describe('#login()', function() {
    it('should login successfuly', function() {
      return netflixLogin.login('johnf@inodes.org', 'secret').then(function(data) {
        expect(data).to.be.an('object');
        expect(data).to.have.property('netflixId').equal(netflixId);
        expect(data).to.have.property('secureNetflixId').equal(secureNetflixId);
        expect(data).to.have.property('esn').equal(esn);
      });
    });
  });

  describe('#expired()', function() {
    it('should return true if no cookies', function() {
      var expired = netflixLogin.expired();
      expect(expired).to.equal(true);
    });
  });
});


test/login.js
import { expect } from 'chai';
import nock from 'nock';

import netflixLogin from '../src/login';

// Setup nock
nock.disableNetConnect();
// nock.enableNetConnect();
// nock.recorder.rec();

const username = 'johnf@inodes.org';
const password = 'secret';
const authURL = 'authURL=1468047668683.authURL';
const netflixId = 'netflixCookie';
const secureNetflixId = 'secureNetflixCookie';
const esn = 'ESN';

const nockOptions = {
  allowUnmocked: false,
};

// getAuthURL
nock('https://www.netflix.com', nockOptions)
  .get('/Login')
  .reply(302, '', { location: 'https://www.netflix.com/gb/Login' });

nock('https://www.netflix.com', nockOptions)
  .get('/gb/Login')
  .reply(200, `name="authURL" value="${authURL}"`);

// postLogin
const body = `authURL=${encodeURIComponent(authURL)}
&email=${encodeURIComponent(username)}
&password=${password}
&rememberMeCheckbox=true
&flow=websiteSignUp
&mode=login
&action=loginAction
&withFields=email%2Cpassword%2CrememberMe%2CnextPage
&nextPage=`;

nock('https://www.netflix.com')
  .post('/Login', body)
  .reply(302, '', {
    location: 'https://www.netflix.com/gb/',
    'set-cookie': [
      `NetflixId=${netflixId}; Domain=.netflix.com; Path=/; Expires=Sun, 09 Jul 2017 12:44:54 GMT; HttpOnly`,
      `SecureNetflixId=${secureNetflixId}; Domain=.netflix.com; Path=/; Expires=Fri, 01 Jan 2038 00:00:00 GMT; HttpOnly; Secure`,
    ],
  });

// getLogin
nock('https://www.netflix.com', nockOptions)
  .get('/Login')
  .reply(302, '', { location: 'https://www.netflix.com/gb/Login' });

nock('https://www.netflix.com', nockOptions)
  .get('/gb/Login')
  .reply(302, '', { location: 'https://www.netflix.com/browse' });

nock('https://www.netflix.com', nockOptions)
  .get('/browse')
  .reply(200, `"esn":"${esn}"`);

describe('Netflix Login', () => {
  beforeEach(() => {
    netflixLogin.privateReset();
  });

  describe('#login()', () => {
    it('should login successfuly', () => (
      netflixLogin.login('johnf@inodes.org', 'secret').then((data) => {
        expect(data).to.be.an('object');
        expect(data).to.have.property('netflixId').equal(netflixId);
        expect(data).to.have.property('secureNetflixId').equal(secureNetflixId);
        expect(data).to.have.property('esn').equal(esn);
      })
    ));
  });

  describe('#expired()', () => {
    it('should return true if no cookies', () => {
      const expired = netflixLogin.expired();
      expect(expired).to.equal(true);
    });
  });
});


test/crypto.js
// import { expect } from 'chai';
import nock from 'nock';

// import netflixCrypto from '../src/crypto';

// Setup nock
nock.disableNetConnect();

describe('Netflix Crypto', () => {
  /*
  describe('#_importNetflixKey()', () => {
    it('should decode the public key', () => {
      const netflixPublicKey = netflixCrypto.importNetflixKey();
      expect(netflixPublicKey).to.be.an('object');
      expect(netflixPublicKey).to.have.property('n');
    });
  });

  describe('#_generateRSAKeyPair()', function moo() {
    this.timeout(10000);

    it('should generate an RSA key pair', () => (
      netflixCrypto.generateRSAKeyPair().then((rsaKeyPair) => {
        expect(rsaKeyPair).to.be.an('object');
        expect(rsaKeyPair).to.have.property('privateKey');
        expect(rsaKeyPair).to.have.property('publicKey');
        expect(rsaKeyPair).to.have.property('publicKeyPem');
      })
    ));
  });
  */
});


src/login.js
import fs from 'fs';
import request from 'request-promise';
import tough from 'tough-cookie';

import Purdy from 'purdy';
import Debug from 'debug';

const debug = Debug('netflix-login');

const baseOptions = {
  uri: 'https://www.netflix.com/Login',
  headers: {
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/46.0.2486.0 Safari/537.36 Edge/13.10586',
  },
  jar: request.jar(),
};

const saveToCache = (authData, filename) => {
  debug('saving to cache');

  const data = {
    // FIXME why are we accessing something private?
    cookieJSON: authData.cookieJar._jar.toJSON(), // eslint-disable-line no-underscore-dangle
    esn: authData.esn,
  };
  const dataJSON = JSON.stringify(data);

  fs.writeFileSync(filename, dataJSON);
};

const loadFromCache = (filename) => {
  debug('loading from cache');

  const dataJSON = fs.readFileSync(filename);
  const data = JSON.parse(dataJSON);

  const cookieJar = new tough.CookieJar.fromJSON(data.cookieJSON); // eslint-disable-line new-cap
  const authData = {
    cookieJar: request.jar(cookieJar.store),
    esn: data.esn,
  };

  return authData;
};

const cookieJar = () => baseOptions.jar;

const cookies = () => cookieJar().getCookies(baseOptions.uri);

const cookiesMissing = () => {
  let result = 0;

  cookies().forEach((cookie) => {
    if (cookie.key === 'NetflixId' || cookie.key === 'SecureNetflixId') {
      result += 1;
    }
  });

  return result !== 2;
};

const getAuthURL = () => (
    request(baseOptions).then((body) => {
      const authURL = body.match(/name="authURL" value="(.+?)"/)[1];
      debug(`authURL: ${authURL}`);

      return authURL;
    })
    .catch((err) => {
      throw new Error(`netflix-login: Couldn't retrieve authURL ${err}`);
    })
);

const postLogin = (authURL, email, password) => {
  const options = {
    ...baseOptions,
    method: 'POST',
    simple: false,
    resolveWithFullResponse: true,
    form: {
      authURL,
      email,
      password,
      rememberMeCheckbox: 'true',
      flow: 'websiteSignUp',
      mode: 'login',
      action: 'loginAction',
      withFields: 'email,password,rememberMe,nextPage',
      nextPage: '',
    },
  };

  return request(options)
    .then((response) => {
      if (response.statusCode !== 302) {
        throw new Error(`netflix-login: Login POST didn't redirect. Bad password? status code: ${response.statusCode}`);
      }

      const location = response.headers.location;
      debug(`redirected to ${location}`);

      return location;
    })
  .catch((err) => {
    throw new Error(`netflix-login: postLogin failed ${Purdy.stringify(err)}`);
  });
};

const getLogin = (/* location */) => (
  request(baseOptions)
    .then((body) => {
      const authData = {
        cookieJar: cookieJar(),
      };

      cookies().forEach((cookie) => {
        if (cookie.key === 'NetflixId') {
          authData.netflixId = cookie.value;
        }
        if (cookie.key === 'SecureNetflixId') {
          authData.secureNetflixId = cookie.value;
        }
      });

      authData.esn = body.match(/"esn":"([^"]+)"/)[1];

      debug(`authData: ${Purdy.stringify(authData)}`);

      return authData;
    })
    .catch((err) => {
      throw new Error(`netflix-login: getLogin failed ${err.stack}`);
    })
);

class NetflixLogin {
  static login(email, password, options = {}) {
    const authDataFilename = `${options.cachePath}/authData.json`;

    if (options.useCache && fs.existsSync(authDataFilename)) {
      return Promise.resolve(loadFromCache(authDataFilename));
    }

    return getAuthURL()
      .then(authURL => postLogin(authURL, email, password))
      .then(location => getLogin(location))
      .then((authData) => {
        if (options.useCache) {
          saveToCache(authData, authDataFilename);
        }

        return authData;
      });
  }

  static expired() {
    let result = false;

    if (cookiesMissing()) {
      return true;
    }

    cookies().forEach((cookie) => {
      if (cookie.key !== 'NetflixId' && cookie.key !== 'SecureNetflixId') {
        return;
      }

      if (cookie.TTL() === 0) {
        result = true;
      }
    });

    return result;
  }

  static privateReset() {
    baseOptions.jar = request.jar();
  }
}

export default NetflixLogin;


index.js
"use strict";

var fs = require('fs');
var request = require('request-promise');
var tough = require('tough-cookie');

var Purdy = require('purdy');
var debug = require('debug')('netflix-login');


module.exports = {
  _baseOptions: {
    uri: 'https://www.netflix.com/Login',
    headers: {
      'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/46.0.2486.0 Safari/537.36 Edge/13.10586',
    },
    jar: request.jar()
  },

  _authDataFilename: 'authData.json',

  login: function(email, password, options) {
    var self = this;

    options = options || {};

    if (options.useCache && fs.existsSync(self._authDataFilename)) {
      return self._loadFromCache();
    }

    return self._getAuthURL().then(function(authURL) {
      return self._postLogin(authURL, email, password);
    }).then(function(location) {
      return self._getLogin(location);
    }).then(function(authData) {
      if (options.useCache) {
        return self._saveToCache(authData);
      }
      else {
        return authData;
      }
    });
  },

  expired: function() {
    var result = false;

    if (this._cookiesMissing()) {
      return true;
    }

    var cookies = this._cookies();
    console.log(cookies);
    for(let cookie of cookies) {
      if (cookie.key !== 'NetflixId' && cookie.key !== 'SecureNetflixId') {
        continue;
      }
      if (cookie.TTL() === 0) {
        result = true;
      }
    }

    return result;
  },

  _saveToCache: function(authData) {
    debug('saving to cache');

    var self = this;

    var data = {
      cookieJSON: authData.cookieJar._jar.toJSON(),
      esn: authData.esn
    };
    var dataJSON = JSON.stringify(data);

    return new Promise(function(resolve, reject) {
      fs.writeFile(self._authDataFilename, dataJSON, function(err) {
        if (err) {
          return reject(err);
        }

        resolve(authData);
      });
    });
  },

  _loadFromCache: function() {
    debug('loading from cache');
    var self = this;

    return new Promise(function(resolve, reject) {
      fs.readFile(self._authDataFilename, function(err, dataJSON) {
        if (err) {
          return reject(err);
        }

        var data = JSON.parse(dataJSON);

        var cookieJar = new tough.CookieJar.fromJSON(data.cookieJSON);
        var authData = {
          cookieJar: request.jar(cookieJar.store),
          esn: data.esn
        };

        resolve(authData);
      });
    });
  },

  _cookiesMissing: function() {
    var result = 0;

    var cookies = this._cookies();
    for(let cookie of cookies) {
      if (cookie.key === 'NetflixId' || cookie.key === 'SecureNetflixId') {
        result++;
      }
    }

    if (result === 2) {
      return false;
    }
    else {
      return true;
    }
  },

  _cookieJar: function() {
    return this._baseOptions.jar;
  },

  _cookies: function() {
    return this._cookieJar().getCookies(this._baseOptions.uri);
  },

  _reset: function() {
    this._baseOptions.jar = request.jar();
  },

  _getAuthURL: function() {
    return request(this._baseOptions).then(function (body) {
      var authURL = body.match(/name="authURL" value="(.+?)"/)[1];
      debug('authURL: ' + authURL);

      return authURL;
    }) .catch(function (err) {
      throw new Error('netflix-login: Couldn\'t retrieve authURL' + err);
    });
  },

  _postLogin: function(authURL, email, password) {
    var myOptions = {
      method: 'POST',
      simple: false,
      resolveWithFullResponse: true,
      form: {
        authURL: authURL,
        email: email,
        password: password,
        rememberMeCheckbox: 'true',
        flow: 'websiteSignUp',
        mode: 'login',
        action: 'loginAction',
        withFields: 'email,password,rememberMe,nextPage',
        nextPage: ''
      },
    };

    var options = Object.assign({}, this._baseOptions, myOptions);

    return request(options).then(function (response) {
      if (response.statusCode !== 302) {
        throw new Error('netflix-login: Login POST didn\'t redirect. status code: ' + response.statusCode);
      }

      var location = response.headers.location;
      debug('redirected to ' + location);

      return location;
    }).catch(function (err) {
      throw new Error('netflix-login: postLogin failed ' + Purdy.stringify(err));
    });
  },

  _getLogin: function(location) {
    var self = this;

    return request(this._baseOptions).then(function (body) {
      var cookies = self._cookies();
      var authData = {
        cookieJar: self._cookieJar()
      };

      for(let cookie of cookies) {
        if (cookie.key === 'NetflixId') {
          authData.netflixId = cookie.value;
        }
        if (cookie.key === 'SecureNetflixId') {
          authData.secureNetflixId = cookie.value;
        }
      }

      authData.esn = body.match(/"esn":"([^"]+)"/)[1];

      debug('authData: ' + Purdy.stringify(authData));

      return authData;
    }).catch(function (err) {
      throw new Error('netflix-login: getLogin failed ' + err.stack);
    });
  },
};
